src/Microsoft.AspNet.Server.Kestrel/Infrastructure/KestrelThread.cs

```diff
@@ -226,11 +206,17 @@ public void Send(Action<ListenerSecondary> callback, ListenerSecondary state)
 
         private void PostCloseHandle(Action<IntPtr> callback, IntPtr handle)
         {
-            lock (_workSync)
+            _closeHandleQueue.Enqueue(new CloseHandle { Callback = callback, Handle = handle });
+            WakeUpLoop();
+        }
+
+        private void WakeUpLoop()
+        {
+            if (_loopIdle)
             {
-                _closeHandleAdding.Enqueue(new CloseHandle { Callback = callback, Handle = handle });
+                _loopIdle = false;
+                _post.Send();
```

Comment:
I'm pretty sure there is a race here.

What if we don't call `uv_async_send` because the loop is apparently not idle when in fact the loop is merely finishing up? It's possible that `_workQueue.IsEmpty` returns true, but before `_loopIdle` is actually set to true, new work is enqueued and `WakeUpLoop` is called.